### PR TITLE
Rate limiting, error message types, and new standardized chat and message types

### PR DESCRIPTION
**Pull request by seanoliver**

### CODE REVIEW

**Comment by seanoliver** [comment]
Two quick notes on the error messages:
- Deployment is still failing due to type errors I'm only getting on the pre commit hook, many of which involve the Supabase code which I'm not as familiar with. We might want to chat about the best way to handle this. I can go in and resolve it but it might take me some time while I get up to speed.
- The error messages are working but as discussed in discord, we're having to hack it a little by sending the content in a 200 request body. I think this can be improved over time, but seems to be working quite well for now and we can also use it to support other error types from the server.

File: app/(main)/chat-layout.tsx [hunk 1]
```diff
@@ -348,7 +348,7 @@ const Sidebar = ({
                       chat && (
                         <SidebarItem
                           onClick={() => setOpen(false)}
-                          key={chat.id}
+                          key={`${chat.id}-${i}`}
```
Comment (swyxio) [inline app/(main)/chat-layout.tsx:351]: thank you

File: app/api/chat/route.ts [hunk 2]
```diff
@@ -251,63 +257,10 @@ export async function POST(req: Request) {
 
       // Insert chat into database.
       await supabase.from('chats').upsert({ id, payload }).throwOnError()
+      console.log('route.ts > payload, insert to db', payload)
     }
-
-    //   experimental_onFunctionCall: async (
-    //     { name, arguments: args },
-    //     createFunctionCallMessages
```
Comment (swyxio) [inline app/api/chat/route.ts:258]: satisfying

File: components/chat-message.tsx [hunk 3]
```diff
@@ -104,7 +106,34 @@ const RenderFunctionMessage = ({ message }: ChatMessageProps) => {
...
+type MessageAuthor = 'user' | 'assistant' | 'fnCall' | 'fnResponse' | 'error'
+
 export function ChatMessage({ message, ...props }: ChatMessageProps) {
+  const [messageAuthor, setMessageAuthor] = useState<MessageAuthor>()
+  const [content, setContent] = useState<string>('')
+  const [messageIcon, setMessageIcon] = useState<JSX.Element>()
+
+  const getAuthorType = (message: Message) => {
+    if (message.role === 'assistant') {
+      if (message.name === 'rate-limit') return 'error'
```
Comment (swyxio) [inline components/chat-message.tsx:118]: ah so we abuse the name field while as assistant? very interesting. we could do more with this.

Comment (swyxio) [reply components/chat-message.tsx:118]: on second thought, lets not. someone else might depend on the `name` field assuming it acts how openai behaves. lets make a new field for stuff we add, e.g.

image

**Comment by swyxio** [comment]
image

the dark mode card looks buggy

**Comment by seanoliver** [comment]
> image
> 
> the dark mode card looks buggy

Very weird — Tailwind should be detecting dark mode and switching the bg-color. The styles are in `chat-message.tsx`. @alecdewitz are you seeing this issue too?

**Pull request merged**